### PR TITLE
Code coverage implementation

### PR DESCRIPTION
**Pull request by IsaacAlao**
### This PR will generate a code coverage report | #70 

The following will be generated/altered:
 - commit status (will display code-coverage percentage).
 - code coverage badge (located in ``./src/README.md`` and will update accordingly after the action runs successfully). 
   - e.g. :
   

 
Why is this needed?:
- To quantify various code being tested
- To identify any potential issues

### CODE REVIEW

**Comment by IsaacAlao** [comment]
Before approving this PR please note that in the ``integration.yml`` file, the ``BRANCH_NAME`` needs to be changed to 'dev'.

Thank you.

**Comment by teamrocket77** [comment]
Fix linting issues, in general

**Comment by teamrocket77** [comment]
Also fix the dependency issues

File: .github/workflows/integration.yml [hunk 1]
```diff
@@ -10,6 +10,50 @@ on:
...
+
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: Cypress run
+        working-directory: src
+        run: |
+          npm ci
+          npm run e2e:headless 
+        #uses: cypress-io/github-action@v5
```
Comment (teamrocket77) [inline .github/workflows/integration.yml:27]: Why's all of this commented out if we don't need it remove it

File: src/cypress.config.js [hunk 2]
```diff
@@ -1,8 +1,33 @@
+// import { defineConfig } from '@cypress/code-coverage'
+
+// export default defineConfig({
+//   e2e: {
+//     setupNodeEvents(on, config) {
+//       // implement node event listeners here
```
Comment (teamrocket77) [inline src/cypress.config.js:6]: Same point here we don't need commented out code

File: src/cypress.config.js [hunk 3]
```diff
@@ -1,8 +1,33 @@
...
+//       on('file:preprocessor', require('@cypress/code-coverage/use-babelrc'));
+//       return config;
+//     },
+//   }
+// })
+
 const { defineConfig } = require('cypress')
 
 module.exports = defineConfig({
-  video : false,
```
Comment (teamrocket77) [inline src/cypress.config.js:4]: We need this to be false, this really Impacts GitHub action time. We can revisit this at a later date with more options if we need

File: src/cypress/integration/example.spec.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+// Example Test.
```
Comment (teamrocket77) [inline src/cypress/integration/example.spec.js:1]: What's the purpose of this test?
Are we able to get it running without this?

File: src/cypress/support/commands.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
+// ***********************************************
+
+// This example commands.js shows you how to
```
Comment (teamrocket77) [inline src/cypress/support/commands.js:3]: Why do we have this file

**Review by teamrocket77 — CHANGES_REQUESTED** [review]
Address changes for favor

**Comment by teamrocket77** [comment]
Integration test run but they break, because you have an error. Why doesn't container build now? 
Check out the GitHub Action logs

**Review by IsaacAlao — COMMENTED** [review]
Fixed.

File: .github/workflows/integration.yml (hunk 1 above)
Comment (IsaacAlao) [reply .github/workflows/integration.yml]: Sorry about that!

File: src/package.json [hunk 6]
```diff
@@ -8,7 +8,10 @@
     "start": "next start -H 0.0.0.0 -p 8080",
     "lint": "next lint",
     "lint-output": "npx eslint --ext js,jsx -c .eslintrc.json --format json -o ../eslint_report.json .",
-    "lint-fix": "npx eslint --ext js,jsx -c .eslintrc.json --fix --ext js.jsx ."
+    "lint-fix": "npx eslint --ext js,jsx -c .eslintrc.json --fix --ext js.jsx .",
+    "cypress:headless": "cypress run",
+    "coverage": "nyc report --reporter=text-summary",
```
Comment (Sanjay-B) [inline src/package.json:13]: Rename this to ``code_coverage``. ``coverage`` is too ambiguous.

File: src/package.json [hunk 7]
```diff
@@ -18,16 +21,25 @@
...
   },
   "devDependencies": {
+    "nyc": "^15.1.0",
+    "@babel/core": "^7.20.12",
+    "@babel/preset-env": "^7.20.2", 
+    "babel-plugin-istanbul": "^6.1.1",
+    "cypress": "^11.2.0",
+    "check-code-coverage": "^1.10.4",
+    "@cypress/code-coverage": "^3.10.0", 
+    "start-server-and-test": "^1.15.3",
```
Comment (Sanjay-B) [inline src/package.json:41]: What is this?

**Review by Sanjay-B — CHANGES_REQUESTED** [review]
LGTM. Clarify some stuff.

File: src/package.json (hunk 7 above)
Comment (IsaacAlao) [reply src/package.json:41]: cypress: what we use to run e2e (end-to-end) tests or component tests
@cypress/code-coverage: saves code coverage reports
check-code-coverage: tools for displaying and presenting code coverage reports
start-server-and-test: starts the webserver and runs the test command

**Review by Sanjay-B — APPROVED** [review]
LGTM.

File: src/package.json (hunk 7 above)
Comment (Sanjay-B) [reply src/package.json:41]: Alright, thanks for clarifying.

**Comment by teamrocket77** [comment]
@IsaacAlao last I checked this PR was not working has anything changed? Will mark as requested changes unless answer is yes lol

File: .github/workflows/integration.yml [hunk 8]
```diff
@@ -10,6 +10,42 @@
...
 
 jobs:
+
+  integration_test:
+    name: Integration Test
+    runs-on: ubuntu-latest
+    steps:
+
+      - name: Checkout
+        uses: actions/checkout@v3
```
[the remark this answers is not in the document]
Comment (teamrocket77) [reply .github/workflows/integration.yml:21]: this should be pinned via sha as seen in other [actions](https://github.com/KSU-Google-DSC/Webapp/blob/dev/.github/workflows/integration.yml)

**Review by teamrocket77 — CHANGES_REQUESTED** [review]
See [me](https://github.com/KSU-Google-DSC/Webapp/pull/89#issuecomment-1416930091)

**Review by IsaacAlao — COMMENTED** [review]
Changed.

**Comment by IsaacAlao** [comment]
@teamrocket77 The only problem with the PR is the docker build job (which I didn't modify and it fails because of some webpack error). I'll investigate the issue sometime later.

File: .github/workflows/integration.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (IsaacAlao) [reply .github/workflows/integration.yml:21]: Will work on that.

**Review by IsaacAlao — COMMENTED** [review]
👍

File: src/cypress/support/index.js [hunk 9]
```diff
@@ -0,0 +1 @@
+import '@cypress/code-coverage/support'
```
Comment (teamrocket77) [inline src/cypress/support/index.js:1]: I know it seemed reasonable, but why are is this present in this file and the e2e.js file? I know that the process was troubleshooting trying to get it working, but are we able to run the code coverage job without it?

File: src/cypress.config.js [hunk 10]
```diff
@@ -1,8 +1,21 @@
...
+    codeCoverage: {
+      url: "/api/__coverage__",
+      exclude: "cypress/**/*.*",
+    }
+  },
   e2e: {
-    supportFile : false,
+    baseUrl: "http://localhost:3000",
+    setupNodeEvents(on, config) {
+      require("@cypress/code-coverage/task")(on, config);
```
Comment (teamrocket77) [inline src/cypress.config.js:17]: In one of the previous commits this pointed to a support file, do we need this line? Do we need the support files?

File: src/package.json [hunk 11]
```diff
@@ -4,33 +4,44 @@
...
     "dev": "next dev",
+    "dev:instrumented": " INSTRUMENT_CODE=1 next dev",
     "build": "next build",
     "start": "next start -H 0.0.0.0 -p 8080",
     "lint": "next lint",
     "lint-output": "npx eslint --ext js,jsx -c .eslintrc.json --format json -o ../eslint_report.json .",
     "lint-fix": "npx eslint --ext js,jsx -c .eslintrc.json --fix --ext js.jsx ."
   },
   "dependencies": {
+    "@babel/core": "^7.20.12",
```
Comment (teamrocket77) [inline src/package.json:15]: Looking back we might be able to get rid of all of these babel plugins, I added them in haste looking to solve the cannot find module issue, if we don't need it let's trim it. I will build locally and see. Check back for more info

**Review by teamrocket77 — CHANGES_REQUESTED** [review]

File: src/package.json (hunk 11 above)
Comment (teamrocket77) [reply src/package.json:15]: It builds locally with without the required files. 
[docker.log](https://github.com/KSU-Google-DSC/Webapp/files/10682257/docker.log)
In addition to that, we are still able to run the code: coverage process, could you try removing these and pushing?

File: src/cypress.config.js (hunk 10 above)
Comment (IsaacAlao) [reply src/cypress.config.js:17]: That line is needed. Support files can be used before spec files get loaded, but I'm not sure we have a use for them now, at the moment.

File: src/cypress/support/index.js (hunk 9 above)
Comment (IsaacAlao) [reply src/cypress/support/index.js:1]: There's really no use for both files right at this time, but it is convenient to have them there if we plan to use them in the near future.

Comment (teamrocket77) [reply src/cypress/support/index.js:1]: @IsaacAlao, Lets remove the things that we don't need right now

Comment (IsaacAlao) [reply src/cypress/support/index.js:1]: And the support directory alongside the files within it are needed to generate the `coverage` and `.nyc_output` directories: 
(scroll down to **E2E code coverage**)
https://docs.cypress.io/guides/tooling/code-coverage

File: src/cypress.config.js (hunk 10 above)
Comment (IsaacAlao) [reply src/cypress.config.js:17]: Actually that line is not needed since it prevents cypress from generating `coverage` and `.nyc_output` in `src`.

**Review by teamrocket77 — APPROVED** [review]

**Pull request merged**